### DISCUSSION TITLE
Scaling

### DISCUSSION THREAD

**Discussion by ageljonathan** [teamalgoritma/community#452] [Q&A] [answered]
Selamat sore
Saya ingin bertanya
Saya sedang mengerjakan lbb regression model dgn case memprediksi harga jual rumah
Saya sudah melakukan cross validation dan scaling dgn cara seperti pd gambar
Bagaimana saya bisa menambahkan

center = attr(wbcd_train_xs,"scaled:center")
scale = attr(wbcd_train_xs,"scaled:scale")

Pd scale data test dgn cara seperti tsb
Terima kasih

![Capture](https://user-images.githubusercontent.com/112919414/202148275-a86b7c97-6a5d-4dbe-86c5-44e8ff580582.JPG)

**Comment by ageljonathan** [comment]
Sori mksd saya

center = attr(housing_train,"scaled:center")
scale = attr(housing_train,"scaled:scale")

Seperti yg diajarkan mas Handoyo kmrn

**Reply by divakartika** [reply]
Dear @ageljonathan 

Terima kasih atas pertanyaan yang disampaikan, pertanyaan Anda sudah kami terima dan sedang kami carikan solusi untuk hal ini. Dapat dibantu untuk kirimkan link ke datasetnya untuk kami coba jalankan?

**Reply by ageljonathan** [reply]
https://www.kaggle.com/datasets/kunwarakash/chennai-housing-sales-price

thanks ka

**Reply by ageljonathan** [reply]
Sudah bisa ka
Ternyata saya perlu mengeluarkan variable target dari data test

![Capture](https://user-images.githubusercontent.com/112919414/202348050-a752ab5f-4beb-406d-957a-6d3759266f9a.JPG)

Berarti tinggal yg ini saya blm dpt solusinya kak

center = attr(wbcd_train_xs,"scaled:center")
scale = attr(wbcd_train_xs,"scaled:scale")

Terima kasih

**Reply by anugrah18003** [reply]
Dear @ageljonathan 

Untuk melakukan scaling berdasarkan inclass yang dicontohkan oleh mas Handoyo bisa menggunakan cara berikut:

> scaling data prediktor
housing_train_s <- scale(housing_train)
housing_test_s <- scale(housing_test , 
                      center = attr(housing_train_s,"scaled:center"), 
                      scale = attr(housing_train_s,"scaled:scale"))

**Comment by ageljonathan** [comment]
Setelah itu saya coba buat model dan prediksi 
Tetapi saat prediksi hasilnya error seperti ini

![Capture2](https://user-images.githubusercontent.com/112919414/202161820-3144317f-810b-4e4d-9e03-e67167d9d273.JPG)
![Capture3](https://user-images.githubusercontent.com/112919414/202161845-ee2da1a3-6168-4b27-99ae-9aa53cac63d4.JPG)

Kira2 knp ya ka

**Comment by idoali** [comment, marked as answer]
Dear Pak Agel,

Terima kasih atas pertanyaannya pak. Error terjadi disana dikarenakan ada level baru bernama KKNAgar di predictor `AREA` yang hanya muncul di data `housing_test` dan tidak pernah muncul di `housing_train`. Solusinya adalah pastikan sebelum cross validation bapak tidak pernah menghapus levels dengan fungsi `droplevels()`. 

Bapak bisa mengecek juga apakah predictor `AREA` yang terdapat pada `housing_train` dan `housing_test` memiliki levels yang sama. Bapak bisa mengeceknya dengan fungsi `levels(housing_train$AREA)` dan `levels(housing_test$AREA)`

**Reply by ageljonathan** [reply]
Baik mas ido
Terima kasih jawabannya

Satu lg mas
Saya mau scaling

![scaling](https://user-images.githubusercontent.com/112919414/202360766-3db68507-50d5-4817-82dc-ea5117d6ad59.JPG)

Pertama saya pisahkan dlu variable target (temasuk variable prediktor yg factor) dan variable prediktor (numerik)
Setelah itu saya scaling

![Scaling 2](https://user-images.githubusercontent.com/112919414/202361081-453fcc55-2915-4929-929b-17f891f1da92.JPG)

Selanjutnya saya coba terapkan ke data test dgn cara yg sama
Bedanya di data test saya menggunakan "center = attr(housing_predict,'scaled:center'), scale =attr(housing_predict,'scaled:scale')" seperti yg diajarkan mas Handoyo

Kira2 udh bener ga mas cara seperti itu
Soalnya saya lihat hasilnya ada yg berbeda

**Reply by ageljonathan** [reply]
Sori hasilnya totally beda bkn ada yg berbeda

**Reply by VictorNugraha** [reply]
Dear @ageljonathan,

Setelah saya mencoba melakukan tahapan dan menerapkan fungsi yang Bapak gunakan, hasil dari _scaling_ pada data train dan test sudah memiliki rentan yang sama. Berikut tahapan dan hasil yang saya lakukan:

Step 1:  Membaca data

`
df <- read.csv("data_input/Chennai houseing sale.csv")
`

Step 2: Cross Validation


```
RNGkind(sample.kind = "Rounding") 
set.seed(123)

index <- sample(x = nrow(df), 
                size = nrow(df) * 0.8) 

df_train <- df[index , ]
df_test <- df[-index ,]
```


Step 3: Memilih kolom numerik 

```
df_train_num <- df_train %>% 
  select(INT_SQFT, DIST_MAINROAD, N_BEDROOM, N_BATHROOM, N_ROOM, QS_ROOMS, QS_BATHROOM, QS_BEDROOM, QS_OVERALL)

df_test_num <- df_test %>% 
  select(INT_SQFT, DIST_MAINROAD, N_BEDROOM, N_BATHROOM, N_ROOM, QS_ROOMS, QS_BATHROOM, QS_BEDROOM, QS_OVERALL)
```

Step 4: Melakukan scalling

```
df_train_num_scaled <- df_train_num %>% 
  scale()

df_test_num_scaled <- scale(x = df_test_num,
                     center = attr(df_train_num_scaled, "scaled:center"),
                     scale = attr(df_train_num_scaled, "scaled:scale"))
```

Step 5: Mengecek hasil scaling

```
df_train_num_scaled[, 1] %>% 
  range()

df_test_num_scaled[, 1] %>% 
  range()
```

![image](https://user-images.githubusercontent.com/28379954/202375125-153c598a-7e5c-46ac-bc47-49a627f856e7.png)

Jika Bapak berkenan, silahkan mencoba lagi tahapan scalling yang dilakukan dengan menerapkan tahapan yang saya lampirkan di atas. 

Demikian informasi yang dapat saya sampaikan, semoga dapat membantu Bapak dalam menyelesaikan permasalahan yang terjadi. Terima kasih.

**Reply by ageljonathan** [reply]
Baik mas Victor
Terima kasih banyak
